### PR TITLE
feat: add support for SQLAlchemy dataclass-mapped tables

### LINKED ISSUE #517: Support for SQLAlchemy dataclass
SQLAlchemy 2.0 introduced PEP 681 support to its integration with dataclasses, but unfortunately this is implemented through a generic type `Mapped[T]` that at the moment is not correctly understood.

```python
from typing import Optional

from serde import serde, from_dict
from sqlalchemy import String, JSON
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(MappedAsDataclass, DeclarativeBase):
    pass


@serde
class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    fullname: Mapped[str] = mapped_column(String(30))
    nickname: Mapped[Optional[str]]
    meta: Mapped[Optional[JSON]] = mapped_column(type_=JSON)


user = from_dict(User, {"id": 1, "name": "john", "fullname": "John Doe"})

```
Leads to an error:
```
Traceback (most recent call last):
  File "[...]/example.py", line 25, in <module>
    user = from_dict(User, {"id": 1, "name": "john", "fullname": "John Doe"})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/lib/python3.12/site-packages/serde/de.py", line 529, in from_dict
    return from_obj(cls, o, named=True, reuse_instances=reuse_instances)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/lib/python3.12/site-packages/serde/de.py", line 497, in from_obj
    raise SerdeError(e) from None
serde.compat.SerdeError: Unsupported type: int
```

### PR DESCRIPTION
**Pull request by barsa-net**
Resolves #517

This hopefully add support for SQLAlchemy dataclass-mapped tables without breaking anything else

The following code should run without errors:
```python
from typing import Optional

import sqlalchemy as sa
from sqlalchemy import String, create_engine, JSON
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Session
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

from serde import serde, from_dict
from serde.json import to_json


class Base(MappedAsDataclass, DeclarativeBase):
    pass


@serde
class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    fullname: Mapped[str] = mapped_column(String(30))
    nickname: Mapped[Optional[str]]
    meta: Mapped[Optional[JSON]] = mapped_column(type_=JSON)


engine = create_engine("sqlite:///:memory:")
Base.metadata.create_all(engine)

user = from_dict(User, {"id": 1, "name": "john", "fullname": "John Doe"})
with Session(engine) as session:
    session.add(user)
    session.commit()

    user = session.scalar(sa.select(User))

to_json(user)
```

### CODE REVIEW

**Comment by barsa-net** [comment]
I added only some simple tests at the moment, but reaching codecov target is unfeasible without bloat testing everything after `except ImportError:`
@yukinarit may I ask for some advices on what other cases than simples one should be covered by tests?

File: tests/test_sqlalchemy.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @serde.serde(**opt)
+    class User(Base):
+        __tablename__ = "user"
+
+        id: Mapped[int] = mapped_column(primary_key=True)
+        name: Mapped[str]
+        fullname: Mapped[str] = mapped_column(String(30))
+
+    user = User(1, "john", "John Doe")
+    assert user == de(User, se(user))
```
Comment (yukinarit) [inline tests/test_sqlalchemy.py:71]: Can you run tests without sqlalchemy?

You can manually uninstall sqlalchemy and run tests here IMO
https://github.com/yukinarit/pyserde/blob/main/.github/workflows/test.yml#L32-L38

**Review by yukinarit — COMMENTED** [review]
Hi @barsa-net 

Thank you for contributing to pyserde! :pray: 

Please do the followings and the PR can be merged.
* Run test without sqlalchemy (See the comments in test_sqlalchemy.py)
* Add simple example (See https://github.com/yukinarit/pyserde/pull/522)
* Add documentation in https://github.com/yukinarit/pyserde/blob/main/docs/en/types.md

File: tests/test_sqlalchemy.py (hunk 1 above)
Comment (barsa-net) [reply tests/test_sqlalchemy.py:71]: Sure can do, but I have a unrelated doubt regarding https://github.com/yukinarit/pyserde/blob/main/.github/workflows/test.yml#L34
All tests are being run through `poetry run ...` while orjson is being installed (seemingly) outside the virtual enviroment with `pip install orjson`, it should be `poetry run pip install orjson` or do I miss something?

**Comment by barsa-net** [comment]
Just to make it clear and avoiding misunderstanding about the scope of this PR, SQLAlchemy is a HUGE project with countless ways to declare tables, as such I plan to add only "basic" support to [declarative dataclass mapping](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping).

As such, I suggest it should be annotated as an "experimental" feature, as some features at the moment don't work, e.g.:
- `@serde(type_check=strict)` does nothing, no errors on wrong types, so I guess no type checking is being performed at Beartype level.
- As SQLAlchemy want fields "annotated" with its `mapped_column()` function I'm unsure if it's even possible to make `serde.core.field()` and `sqlalchemy.orm.mapped_column()` working along each other, that means no field attributes.

File: tests/test_sqlalchemy.py (hunk 1 above)
Comment (yukinarit) [reply tests/test_sqlalchemy.py:70]: Good catch. This line had been added two years ago by [this commit](https://github.com/yukinarit/pyserde/commit/2744675cd5af70426411d2e15aeb536e0c07c02b) but apparently pyserde was never tested with orjson :rofl:

**Review by yukinarit — APPROVED** [review]
LGTM!

**Comment by yukinarit** [comment]
@barsa-net is it ready for merge, or you would be adding some more changes?

**Comment by barsa-net** [comment]
Sorry, I have been busy this days.
I want to add a notice in [Types docs](https://github.com/yukinarit/pyserde/blob/4bb9971fda00f2ebf2ef5fc12964a7a92ca828fb/docs/en/types.md) about that is not _feature complete_ (refer to my previous comment)
Also I'm trying something else but I don't know if it works at this time, I'll get in touch by the end of the week.

**Comment by barsa-net** [comment]
Sorry for the delay, I added an additional test for nested SQLAlchemy models and I made clear in the documentation that the feature has to be considered experimental and as such some features could not work.

I consider the PR "complete", if I manage to get some other feature working together with SQLAlchemy I'll eventually open another PR.

**Pull request merged**